tests/phpunit/integration/Core/Modules/REST_Modules_ControllerTest.php

```diff
@@ -0,0 +1,521 @@
+
+		$request = new WP_REST_Request( 'POST', '/' . REST_Routes::REST_ROOT . '/core/modules/data/recover-module' );
+		$request->set_body_params(
+			array(
+				'data' => array(
+					'slug' => 'search-console',
+				),
+			)
+		);
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertEquals( 'module_not_recoverable', $response->get_data()['code'] );
+		$this->assertEquals( 403, $response->get_status() );
+	}
+
+	public function test_recover_module_rest_endpoint__requires_accessible_module() {
+		$this->enable_feature( 'dashboardSharing' );
+
+		// Make search-console a recoverable module
+		$search_console = $this->modules->get_module( 'search-console' );
+		$search_console->get_settings()->merge(
+			array(
+				'propertyID' => '123456789',
+			)
+		);
+		$test_sharing_settings = array(
+			'search-console' => array(
+				'sharedRoles' => array( 'editor', 'subscriber' ),
+				'management'  => 'owner',
+			),
+		);
+		add_option( 'googlesitekit_dashboard_sharing', $test_sharing_settings );
+
+		$request = new WP_REST_Request( 'POST', '/' . REST_Routes::REST_ROOT . '/core/modules/data/recover-module' );
+		$request->set_body_params(
+			array(
+				'data' => array(
+					'slug' => 'search-console',
+				),
+			)
+		);
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertEquals( 'module_not_accessible', $response->get_data()['code'] );
+		$this->assertEquals( 403, $response->get_status() );
+	}
+
+	public function test_recover_module_rest_endpoint__success() {
+		$this->enable_feature( 'dashboardSharing' );
+
+		// Make search-console a recoverable module
+		$search_console = $this->modules->get_module( 'search-console' );
+		$search_console->get_settings()->merge(
+			array(
+				'propertyID' => '123456789',
+			)
+		);
+		$test_sharing_settings = array(
+			'search-console' => array(
+				'sharedRoles' => array( 'editor', 'subscriber' ),
+				'management'  => 'owner',
+			),
+		);
+		add_option( 'googlesitekit_dashboard_sharing', $test_sharing_settings );
+
+		// Make search-console service requests accessible
+		$search_console->get_client()->setHttpClient( new FakeHttpClient() );
+
+		$request = new WP_REST_Request( 'POST', '/' . REST_Routes::REST_ROOT . '/core/modules/data/recover-module' );
+		$request->set_body_params(
+			array(
+				'data' => array(
+					'slug' => 'search-console',
+				),
+			)
+		);
+		$response = rest_get_server()->dispatch( $request );
+
+		$current_user = wp_get_current_user();
+		$this->assertEquals(
+			array(
+				'ownerID' => $current_user->ID,
+			),
+			$response->get_data()
+		);
+		$this->assertEquals( 200, $response->get_status() );
+	}
+
+	public function test_list_rest_endpoint__get_method() {
+		$request  = new WP_REST_Request( 'GET', '/' . REST_Routes::REST_ROOT . '/core/modules/data/list' );
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertEquals( 200, $response->get_status() );
+		$this->assertNotEmpty( $response->get_data() );
```

Comment:
Can we use a few assertions to ensure the output is in the shape we expect?

Reply:
@aaemnnosttv I've added `test_list_rest_endpoint__shape`. Let me know if that works!